core/common/src/main/java/alluxio/concurrent/CountLatch.java

```diff
@@ -0,0 +1,155 @@
+ * See alluxio.client.fs.FileSystemContextReinitializer for more details.
+ */
+public class CountLatch {
+  /** Represents the ignored arguments for AQS methods. */
+  private static final int IGNORED_ARG = -1;
+  /** Represents the state where {@link #inc()} is blocked. */
+  private static final int BLOCK_INC_STATE = -1;
+
+  /**
+   * Meaning of the AQS state:
+   * 1. when equals BLOCK_INC_STATE, means awaitAndBlockInc has returned, and all further calls to
+   * inc are blocked until unblockInc is called.
+   * 2. when >= 0, means the counter value increased by inc and decreased by dec.
+   */
+  private static final class Sync extends AbstractQueuedSynchronizer {
+    private static final long serialVersionUID = 8553010505281724322L;
+
+    @Override
+    protected boolean tryAcquire(int arg) {
+      while (true) {
+        if (getState() != 0) {
+          return false;
+        }
+        if (compareAndSetState(0, BLOCK_INC_STATE)) {
+          return true;
+        }
+      }
+    }
+
+    @Override
+    protected boolean tryReleaseShared(int arg) {
+      if (getState() != BLOCK_INC_STATE || !compareAndSetState(BLOCK_INC_STATE, 0)) {
+        throw new Error("state must be " + BLOCK_INC_STATE);
+      }
+      return true;
+    }
+
+    @Override
+    protected int tryAcquireShared(int arg) {
+      while (true) {
+        int current = getState();
+        if (current == BLOCK_INC_STATE) {
+          return -1;
+        }
+        if (compareAndSetState(current, current + 1)) {
+          return 1;
+        }
+      }
+    }
+
+    @Override
+    protected boolean tryRelease(int arg) {
+      while (true) {
+        int current = getState();
+        if (current <= 0) {
+          throw new Error("state must be > 0");
+        }
+        int newState = current - 1;
+        if (compareAndSetState(current, newState)) {
+          // Signal only when transitioning to 0.
+          return newState == 0;
+        }
+      }
+    }
+
+    @VisibleForTesting
+    int state() {
+      return getState();
+    }
+  }
+
+  private Sync mSync = new Sync();
+
+  /**
+   * Increases the counter.
+   *
+   * If {@link #awaitAndBlockInc()} returns, this call is blocked until {@link #unblockInc()} is
+   * called.
+   *
+   * @throws InterruptedException when interrupted during being blocked
+   */
+  public void inc() throws InterruptedException {
+    mSync.acquireSharedInterruptibly(IGNORED_ARG);
+  }
+
+  /**
+   * Decreases the counter.
+   *
+   * Should only be called in pair with {@link #inc()}.
+   * If counter will go below zero after this call, throws {@link Error}.
+   * This method is never blocked.
+   */
+  public void dec() {
+    mSync.release(IGNORED_ARG);
+  }
+
+  /**
+   * Blocked during awaiting the counter to become zero.
+   * When it returns, all further calls to {@link #inc()} are blocked.
+   */
+  public void awaitAndBlockInc() {
```

Comment:
NIT: Just await(). The name "Latch", combined with the comment makes it explicit enough.